hazelcast-jet-core/src/test/java/com/hazelcast/jet/core/metrics/MetricsTest.java

```diff
@@ -0,0 +1,279 @@
+
+        assertMetricsProduced(pipeline, "dropped", 2, "total", 5);
+    }
+
+    @Test
+    public void gauge() {
+        pipeline.drawFrom(TestSources.items(0L, 1L, 2L, 3L, 4L))
+                .mapStateful(LongAccumulator::new, (acc, i) -> {
+                    acc.add(i);
+                    Metric metric = Metrics.metric("sum", Unit.COUNT);
+                    metric.set(acc.get());
+                    return acc.get();
+                })
+                .drainTo(Sinks.logger());
+
+        assertMetricsProduced(pipeline, "sum", 10);
+    }
+
+    @Test
+    public void gauge_notUsed() {
+        pipeline.drawFrom(TestSources.items(0L, 1L, 2L, 3L, 4L))
+                .mapStateful(LongAccumulator::new, (acc, i) -> {
+                    acc.add(i);
+                    Metrics.metric("sum", Unit.COUNT);
+                    return acc.get();
+                })
+                .drainTo(Sinks.logger());
+
+        assertMetricsProduced(pipeline, "sum", 0L);
+    }
+
+    @Test
+    public void nonCooperativeProcessor() {
+        DAG dag = new DAG();
+
+        Vertex source = dag.newVertex("source", TestProcessors.ListSource.supplier(asList(1L, 2L, 3L)));
+        Vertex map = dag.newVertex("map", new NonCoopTransformPSupplier((FunctionEx<Long, Long>) l -> {
+            Metrics.metric("mapped").inc();
+            return l * 10L;
+        }));
+        Vertex sink = dag.newVertex("sink", writeListP("results"));
+
+        dag.edge(between(source, map)).edge(between(map, sink));
+
+        assertMetricsProduced(dag, "mapped", 3L);
+        assertEquals(
+                new HashSet<>(Arrays.asList(10L, 20L, 30L)),
+                new HashSet<>(instance.getList("results"))
+        );
+    }
+
+    @Test
+    public void metricsDisabled() {
+        Long[] input = {0L, 1L, 2L, 3L, 4L};
+        pipeline.drawFrom(TestSources.items(input))
+                .map(l -> {
+                    Metrics.metric("mapped").inc();
+                    Metrics.metric("total", Unit.COUNT).set(input.length);
+                    return l;
+                })
+                .drainTo(Sinks.logger());
+
+        Job job = instance.newJob(pipeline, new JobConfig().setMetricsEnabled(false));
+        job.join();
+
+        JobMetrics metrics = job.getMetrics();
+        assertTrue(metrics.get("mapped").isEmpty());
+        assertTrue(metrics.get("total").isEmpty());
+    }
+
+    @Test
+    public void usingContextAsync() {
+        pipeline.drawFrom(TestSources.items(0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11))
+                .addTimestamps(i -> i, 0L)
+                .window(WindowDefinition.tumbling(3))
+                .aggregate(summingLong((ToLongFunctionEx<Integer>) Integer::longValue))
+                .map(WindowResult::result)
+                .filterUsingServiceAsync(
+                        ServiceFactory.withCreateFn(i -> 0L),
+                        (ctx, l) -> {
+                            Metric dropped = Metrics.metric("dropped");
```

Comment:
We should use the thread-safe version here. We could actually stress-test it a bit by having lot more items and check that the total matches.